import grpc
import numpy as np
import time
import hyperspace_pb2
import hyperspace_pb2_grpc

# Config
DIM = 8
HOST = "localhost:50051"

def generate_poincare_point(dim):
    """Generates a random point inside the Poincar√© ball"""
    vec = np.random.uniform(-1, 1, dim)
    norm = np.linalg.norm(vec)
    if norm >= 1:
        vec = vec / (norm + 0.1) # Normalize inside
    return vec.tolist()

def run():
    print(f"Connecting to {HOST}...")
    try:
        channel = grpc.insecure_channel(HOST)
        stub = hyperspace_pb2_grpc.DatabaseStub(channel)
        
        # Check connection implicitly by calling Insert
        pass
    except Exception as e:
        print(f"Connection failed: {e}")
        return

    channel = grpc.insecure_channel(HOST)
    stub = hyperspace_pb2_grpc.DatabaseStub(channel)

    # 1. Insert Data with Metadata
    print("üöÄ Inserting vectors...")
    start = time.time()
    for i in range(100):
        vec = generate_poincare_point(DIM)
        # Half "red", Half "blue"
        category = "red" if i % 2 == 0 else "blue"
        meta = {"category": category}
        
        try:
            stub.Insert(hyperspace_pb2.InsertRequest(vector=vec, metadata=meta))
            if i % 10 == 0: print(f".", end="", flush=True)
        except grpc.RpcError as e:
            print(f"Insert failed: {e}")
            break
    
    elapsed = time.time() - start
    print(f"\n‚úÖ Inserted 100 vectors in {elapsed:.4f}s")
    
    # 2. DELETE Test
    print("\nüóëÔ∏è Deleting Vector ID 0 (which is Red)...")
    try:
        stub.Delete(hyperspace_pb2.DeleteRequest(id=0))
        print("‚úÖ Delete confirmed.")
    except grpc.RpcError as e:
        print(f"‚ùå Delete failed: {e}")

    # 3. Search with Filter
    query_vec = generate_poincare_point(DIM)
    
    print(f"\nüîç Searching nearest neighbors (Filter: category='red')...")
    try:
        # Filter for RED only
        f = {"category": "red"}
        response = stub.Search(hyperspace_pb2.SearchRequest(vector=query_vec, top_k=5, filter=f))
        print("üéØ Search Results (Red Only):")
        for res in response.results:
            print(f"   ID: {res.id}, Dist: {res.distance:.4f}")
            
            if res.id == 0:
                print(f"   ‚ö†Ô∏è FATAL: ID 0 found! Soft Delete FAILED!")
            
            if res.id % 2 != 0:
                print(f"   ‚ö†Ô∏è ERROR: Found ID {res.id} which should be blue (odd)!")
    except grpc.RpcError as e:
        print(f"‚ùå RPC Error: {e.details()}")

    print(f"\nüîç Searching nearest neighbors (Filter: category='blue')...")
    try:
        # Filter for BLUE only
        f = {"category": "blue"}
        response = stub.Search(hyperspace_pb2.SearchRequest(vector=query_vec, top_k=5, filter=f))
        print("üéØ Search Results (Blue Only):")
        for res in response.results:
            print(f"   ID: {res.id}, Dist: {res.distance:.4f}")
            if res.id % 2 == 0:
                print(f"   ‚ö†Ô∏è ERROR: Found ID {res.id} which should be red (even)!")
    except grpc.RpcError as e:
        print(f"‚ùå RPC Error: {e.details()}")

if __name__ == "__main__":
    run()
